### PR TITLE
Add board support for nrf52840_mdk_m2_module

### PR DESCRIPTION
**Pull request by mattande**
Origin: Original

Source code from board/arm/nrf52840_mdk_usb_dongle was adapted for the nrf52840_mdk_m2_module.

Modifications:
 - LED's are on different pins
 - Add QSPI
 - Remove CONFIG_NFCT_PINS_AS_GPIO=y as these pins are connected to a u.FL for NFC antenna

### CODE REVIEW

File: boards/arm/nrf52840_mdk_m2_module/Kconfig.defconfig [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	default 0x1000
+	depends on BOARD_HAS_NRF5_BOOTLOADER && !USE_DT_CODE_PARTITION
+
+if USB_DEVICE_STACK
+
+# Enable UART driver, needed for CDC ACM
+config SERIAL
+	default y
+
+endif # USB_DEVICE_STACK
```
Comment (gmarull) [inline boards/arm/nrf52840_mdk_m2_module/Kconfig.defconfig:37]:
```suggestion
# Enable UART driver, needed for CDC ACM
config SERIAL
	default y if USB_DEVICE_STACK
```

File: boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+			pwms = <&pwm0 29>;
+		};
+		blue_pwm_led: pwm_led_2 {
+			pwms = <&pwm0 31>;
+		};
+	};
+
+    /* The button is not present on the module, though reset is typically
+     * connected on the baseboard to reset the module. */
+	buttons {
```
Comment (gmarull) [inline boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts:59]: please don't mix tabs/spaces

File: boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	};
+
+	/* These aliases are provided for compatibility with samples */
+	aliases {
+		led0 = &led0_red;
+		led1 = &led0_green;
+		led2 = &led0_blue;
+		led0-red   = &led0_red;
+		led0-green = &led0_green;
+		led0-blue  = &led0_blue;
```
Comment (gmarull) [inline boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts:74]:
```suggestion
		led0-red = &led0_red;
		led0-green = &led0_green;
		led0-blue = &led0_blue;
```

File: boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+};
+
+&gpio1 {
+	status = "okay";
+};
+
+&uart0 {
+    /* Line drivers are not present on the module, P0.16(RX) / P0.15(TX)
+     * are defined here for compatibility with the DAPLink/LPC11U35 on
+     * the M.2 Developer Kit */
```
Comment (gmarull) [inline boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts:99]: ditto

File: boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+};
+
+&pwm0 {
+	status = "okay";
+	ch0-pin = <29>;
+	ch0-inverted;
+	ch1-pin = <30>;
+	ch1-inverted;
+	ch2-pin = <31>;
+	ch2-inverted;
```
Comment (gmarull) [inline boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts:114]: all in-tree boards have been migrated to pinctrl. You can refer to #43926 for details on how automate the migration using the porting script.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: boards/arm/nrf52840_mdk_m2_module/Kconfig.defconfig (hunk 1 above)
Comment (mattande) [reply boards/arm/nrf52840_mdk_m2_module/Kconfig.defconfig]: fixed in [d0f9bd7](https://github.com/zephyrproject-rtos/zephyr/pull/43992/commits/d0f9bd7f2a6b1c6a706b80ed9617d1cd5b31e48a)

File: boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts (hunk 2 above)
Comment (mattande) [reply boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts]: fixed in [d0f9bd7](https://github.com/zephyrproject-rtos/zephyr/pull/43992/commits/d0f9bd7f2a6b1c6a706b80ed9617d1cd5b31e48a)

File: boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts (hunk 4 above)
Comment (mattande) [reply boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts]: fixed in [d0f9bd7](https://github.com/zephyrproject-rtos/zephyr/pull/43992/commits/d0f9bd7f2a6b1c6a706b80ed9617d1cd5b31e48a)

File: boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts (hunk 3 above)
Comment (mattande) [reply boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts]: fixed in [d0f9bd7](https://github.com/zephyrproject-rtos/zephyr/pull/43992/commits/d0f9bd7f2a6b1c6a706b80ed9617d1cd5b31e48a)

Comment (mattande) [reply boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts]: fixed in [d0f9bd7](https://github.com/zephyrproject-rtos/zephyr/pull/43992/commits/d0f9bd7f2a6b1c6a706b80ed9617d1cd5b31e48a)

File: boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts (hunk 5 above)
Comment (mattande) [reply boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts]: Updated to pinctrl in [dd6f73e](https://github.com/zephyrproject-rtos/zephyr/pull/43992/commits/dd6f73e0f3c748e64558eb2bc44574e5dcb23429)

File: boards/arm/nrf52840_mdk_m2_module/doc/index.rst [hunk 6]
```diff
@@ -0,0 +1,38 @@
+.. _nrf52840_mdk_m2_module:
+
+nRF52840 MDK M.2 Module
```
Comment (jfischer-no) [inline boards/arm/nrf52840_mdk_m2_module/doc/index.rst:3]: We do not add support for modules, it does not make sense. You can add support for [nrf52840-m2-developer-kit](https://makerdiary.com/products/nrf52840-m2-developer-kit)

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: boards/arm/nrf52840_mdk_m2_module/doc/index.rst (hunk 6 above)
Comment (kdewald) [reply boards/arm/nrf52840_mdk_m2_module/doc/index.rst:3]: @jfischer-no The module is a standalone product that is being actively consumed by the community. The developer kit is a breakout board with no real functionality on its own, the value comes from the M2 card.

I think the Zephyr community would benefit from the developers being open to adding support for these boards, as it's a new form factor that's gaining some traction.

I'm willing to help on making the necessary changes to get this MR landed.

File: boards/arm/nrf52840_mdk_m2_module/Kconfig.defconfig [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+if BOARD_NRF52840_MDK_M2_MODULE
+
+config BOARD
+	default "nrf52840_mdk_m2_module"
+
+# To let the nRF5 bootloader load an application, the application
+# must be linked after Nordic MBR, that is factory-programmed on the board.
+
+# Nordic nRF5 booatloader exists outside of the partitions specified in the
```
Comment (kdewald) [inline boards/arm/nrf52840_mdk_m2_module/Kconfig.defconfig:19]: Typo

File: boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+		led0_blue: led_2 {
+			gpios = <&gpio0 31 GPIO_ACTIVE_LOW>;
+			label = "Blue LED 3";
+		};
+	};
+
+	pwmleds {
+		compatible = "pwm-leds";
+		red_pwm_led: pwm_led_0 {
+			pwms = <&pwm0 30>;
```
Comment (kdewald) [inline boards/arm/nrf52840_mdk_m2_module/nrf52840_mdk_m2_module.dts:48]: I had to add `PWM_MSEC(20) PWM_POLARITY_INVERTED` for all PWM definitions for them to work on a newer Zephyr version.

File: boards/arm/nrf52840_mdk_m2_module/doc/index.rst (hunk 6 above)
Comment (jfischer-no) [reply boards/arm/nrf52840_mdk_m2_module/doc/index.rst:3]: > @jfischer-no The module is a standalone product that is being actively consumed by the community. The developer kit is a breakout board with no real functionality on its own, the value comes from the M2 card.
> 

No, [nrf52840-m2-developer-kit](https://makerdiary.com/products/nrf52840-m2-developer-kit) is a board with a DAPLink  adapter, Arduino headers and USB connector and can be easily integrated into Zephyr. A module is usually the SoC + some minimal circuit and the actual configuration changes depending on which board it is installed on. Of course, you can also approach it differently, insert m2_module as a board (btw, it does not meet the FCC requirements for modules, so I doubt anyone will benefit from it at all) and developer kit as a shield, but I assume that board_runner_args cannot be defined at shield level. Anyway, it is not convincing.

> I think the Zephyr community would benefit from the developers being open to adding support for these boards, as it's a new form factor that's gaining some traction.
> 
> I'm willing to help on making the necessary changes to get this MR landed.

"gaining some traction", come on, please do not exaggerate, https://docs.zephyrproject.org/latest/boards/arm/reel_board/doc/index.html#id4, and that was designed 3-4 years ago.

**Pull request closed without merging**